samples/net/cloud/mqtt_azure/src/main.c

```diff
@@ -305,6 +308,21 @@ static int publish(struct mqtt_client *client, enum mqtt_qos qos)
 	return mqtt_publish(client, &param);
 }
 
+static void poll_timeout(struct k_work *work)
+{
+	int rc;
+
+	rc = wait(0);
+	if (rc <= 0) {
+		goto end;
+	}
+
+	mqtt_input(&client_ctx);
```

Comment:
Doing this periodically would introduce some delay anyway (up to 1 s in this case). I think we should simply allow `poll` doing its job to report any new data as soon as possible. This is how I would see this working:

1. Move the poll/mqtt_input combo at the end of the main, it's not doing anything else anyway:
```
while (true) {
	rc = wait(NET_WAIT_FOREVER);
	if (rc < 0) {
		break;
	}

	if (mqtt_input(&client_ctx) < 0) {
		break;
	}
}
```
2. Remove wait/mqtt_input from `publish_message` just as it is.
3. Optionally verify if `POLLIN` was set before calling `mqtt_input` to avoid unnecessary calls. Ideally, verify also other possible poll events (POLLERR/POLLHUP/POLLNVAL), but I guess this might be out of the scope of this PR.

Reply:
>     1. Move the poll/mqtt_input combo at the end of the main, it's not doing anything else anyway:

It doesn't go there at all, connect_to_cloud_and_publish() has while(true) already and publish is waiting for semaphore. I will look into it.

Reply:
> Doing this periodically would introduce some delay anyway (up to 1 s in this case).

by the way, the delay without these changes were based on publish timeout which is between 10-15 + seconds. Anyway if somebody wants experiment they can modify APP_POLL_MSECS. Hope this is fine. what do you say?

Reply:
I still think that if it's possible to avoid extra delays, we should do it.

I missed the fact that the main thread blocks on a semaphore to delay the call to `publish()`. But since we have a workqueue item for publish anyway, we could simply send the publish message directly from the system workqueue, no need to block the main thread. And instead of calling poll/mqtt_input from the workqueue then, you could call it from main, block on poll and process the input as soon as it's avaialble.

Reply:
> I still think that if it's possible to avoid extra delays, we should do it.

I didn't say that it is impossible.

Reply:
> I missed the fact that the main thread blocks on a semaphore to delay the call to `publish()`. But since we have a workqueue item for publish anyway, we could simply send the publish message directly from the system workqueue,

This was your suggestion, when I send this sample, sending publish messages from workqueue. Not good to use system workqueue to call heavy network operations. So keep the publish in main loop and use semaphore.

Reply:
I needed to refresh my memory :)
Indeed I suggested to move `publish_message` into the main thread, but the main reason for that was the `wait()` call within this function, which would block the function (and thus the workqueue):

Reply:
hmm, ok.